### PR TITLE
Add implicit return to original .ctor block

### LINKED ISSUE #7265: Locals are missing in EE at closing brace of .ctor
Break at the closing brace of ```C..ctor```. The Locals window contains ```this``` but not ```o```.

```
class C
{
    C()
    {
        var o = new object();
    }
    static void Main()
    {
        var c = new C();
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #7265

### CODE REVIEW

**Comment by davkean** [comment]
I don't recognized those failures, do you think they were caused by your change?

**Comment by cston** [comment]
@davkean Yes, the build failures are caused by this change.

**Comment by gafter** [comment]
What is the purpose of this change?

**Comment by cston** [comment]
@gafter Locals in a constructor are out of scope in the EE at the closing brace.

**Comment by cston** [comment]
@gafter, @dotnet/roslyn-compiler, @amcasey

**Comment by cston** [comment]
Fixes #7265

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs [hunk 1]
```diff
@@ -78,6 +80,25 @@ internal class FlowAnalysisPass
...
         }
 
+        private static BoundBlock AppendImplicitReturn(BoundBlock body, bool originalBodyNested, MethodSymbol method, CSharpSyntaxNode syntax)
+        {
+            if (originalBodyNested)
+            {
+                var statements = body.Statements;
+                int n = statements.Length;
+
+                var builder = ArrayBuilder<BoundStatement>.GetInstance();
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs]: `GetInstance(n)`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs [hunk 2]
```diff
@@ -78,6 +80,25 @@ internal class FlowAnalysisPass
             return block;
         }
 
+        private static BoundBlock AppendImplicitReturn(BoundBlock body, bool originalBodyNested, MethodSymbol method, CSharpSyntaxNode syntax)
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs]: I might have put the new flag at the end to make it easier to use named arguments.

**Comment by amcasey** [comment]
:+1:

**Comment by jaredpar** [comment]
How does this change fix that particular bug thuogh.  It's not obivous at a glance.

**Comment by amcasey** [comment]
@jaredpar The implicit return statement gets a sequence point.

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs [hunk 3]
```diff
@@ -78,6 +80,25 @@ internal class FlowAnalysisPass
...
+        {
+            if (originalBodyNested)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs]: Could we instead check WasCompilerGenerated flag on the BoundBlock node?

Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs]: WasCompilerGenerated will be set in the case where there is a constructor initializer but no original body as well. Passing an explicit flag seemed less fragile.

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs]: Changed.

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs]: Done.

**Comment by AlekseyTs** [comment]
LGTM

**Pull request merged**